slotsmanager.c

```diff
@@ -335,6 +346,72 @@ slots_gc_slots()
     return 0;
 }
 
+static int
+order_by_type_and_key(const void *a, const void *b)
+{
+    const struct particle_data * pa  = (const struct particle_data *) a;
+    const struct particle_data * pb  = (const struct particle_data *) b;
+
+    if(pa->IsGarbage && !pb->IsGarbage)
+        return +1;
+    if(!pa->IsGarbage && pb->IsGarbage)
+        return -1;
+    if(pa->Type < pb->Type)
+        return -1;
+    if(pa->Type > pb->Type)
+        return +1;
+    if(pa->Key < pb->Key)
+        return -1;
+    if(pa->Key > pb->Key)
+        return +1;
+
+    return 0;
+}
+
+/* Sort the particles and their slots by type and peano order.
+ * This does a gc by sorting the Garbage to the end of the array and then trimming.
+ * It is a different algorithm to slots_gc, somewhat slower,
+ * but delivers a spatially compact sort. It always compacts the slots*/
+void
+slots_gc_sorted()
+{
+    int ptype;
+    /* Resort the particles such that those of the same type and key are close by.
+     * The locality is broken by the exchange. */
+    qsort_openmp(P, NumPart, sizeof(struct particle_data), order_by_type_and_key);
+
+    /*Reduce NumPart*/
+    while(P[NumPart-1].IsGarbage) {
+        NumPart--;
+    }
+
+    /*Set up ReverseLink*/
+    slots_gc_mark();
+
+    for(ptype = 0; ptype < 6; ptype++) {
+        if(!SLOTS_ENABLED(ptype))
+            continue;
+        /* sort the used ones
+         * by their location in the P array */
+        qsort_openmp(SlotsManager->info[ptype].ptr,
+                 SlotsManager->info[ptype].size,
+                 SlotsManager->info[ptype].elsize,
+                 slot_cmp_reverse_link);
+
+        /*Reduce slots used*/
+        while(BASESLOT_PI(SlotsManager->info[ptype].size-1, ptype)->IsGarbage) {
+            SlotsManager->info[ptype].size--;
+        }
+        slots_gc_collect(ptype);
+    }
+#ifdef DEBUG
+    slots_check_id_consistency();
+#endif
+
+    /*Rebuild the tree if still allocated*/
+    EISlotsAfterGC event = {{1}};
```

Comment:
If we no longer need the event -- then shall remove it as well.. the comment looked out dated.

Reply:
Ok, done.